### PR TITLE
Resolving issue #1745. Added cluster option for SNI data

### PR DESCRIPTION
**Pull request by KlimentSerafimov**
#### Summary

Added ability to pass cluster name as an option argument (..&options=cluster:<project name>..) in case SNI data is missing.
Resolving issue #1745 

#### Rationale

Since SNI is a recent postgres feature, some posgres clients might not be updated, and since our proxy relies on the SNI data for routing, we need to prove an alternative way for users to pass this information. For that purpose we are added logic to use the cluster parameter from options argument as the SNI data

#### Behavior

After starting the proxy
`./target/debug/proxy --auth-backend=console -c server.crt -k server.key`
running (with sslsni=0)
`psql 'postgres://my-cluster-123.localtest.me:4432/main?sslmode=require&sslsni=0'`
results in (notice new error message)
> got message: StartupMessage { major_version: 3, minor_version: 0, params: {"": "", "user": "klimentserafimov", "application_name": "psql", "database": "main", "client_encoding": "UTF8"} }
error: SNI info is missing. EITHER upgrade the postgres client library OR pass ..&options=cluster:<project name>.. parameter

Running (with ..&options=project:my-cluster-123)
`psql 'postgres://my-cluster-123.localtest.me:4432/main?sslmode=require&sslsni=0&options=project:my-cluster-123'`
results in (notice cluster name in StartupMessage)

> got message: StartupMessage { major_version: 3, minor_version: 0, params: {"user": "klimentserafimov", "application_name": "psql", "database": "main", "client_encoding": "UTF8", "": "", "project": "my-cluster-123"} }
cplane request: http://localhost:3000/authenticate_proxy_request//proxy_get_role_secret?project=my-cluster-123&role=klimentserafimov
error: error sending request for url (http://localhost:3000/authenticate_proxy_request//proxy_get_role_secret?project=my-cluster-123&role=klimentserafimov): error trying to connect: tcp connect error: Connection refused (os error 61)

Which is exactly what we want, because when we run without sslsni=0 and without .&options=project:my-cluster-123:
`psql 'postgres://my-cluster-123.localtest.me:4432/main?sslmode=require'`
results in

> got message: StartupMessage { major_version: 3, minor_version: 0, params: {"user": "klimentserafimov", "application_name": "psql", "client_encoding": "UTF8", "": "", "database": "main"} }
cplane request: http://localhost:3000/authenticate_proxy_request//proxy_get_role_secret?project=my-cluster-123&role=klimentserafimov
error: error sending request for url (http://localhost:3000/authenticate_proxy_request//proxy_get_role_secret?project=my-cluster-123&role=klimentserafimov): error trying to connect: tcp connect error: Connection refused (os error 61)

### CODE REVIEW

**Comment by funbringer** [comment]
@KlimentSerafimov sorry for disrupting your workflow, I've just merged https://github.com/neondatabase/neon/pull/1802 which re-organizes some modules and fixes a couple of problems with error reporting. I put logic for handling project name (via sni) [here](https://github.com/neondatabase/neon/blob/b3ec6e0661e1f08beb1cd08b265cc64af0cd4035/proxy/src/auth/credentials.rs#L51), so I guess you could move your [changes](https://github.com/neondatabase/neon/blob/2860e71b9a24c89a766b027e3d1fbe1f4ce966ab/proxy/src/auth_backend/console.rs#L172-L186) to that function.

File: libs/utils/src/pq_proto.rs [hunk 1]
```diff
@@ -269,11 +269,15 @@ impl FeStartupPacket {
...
-                                let nameval: Vec<&str> = cmdopt.split('=').collect();
+                            //parsing options arguments "..&options=<var>:<val>,.."
+                            //extended example and set of options:
+                            //https://github.com/neondatabase/neon/blob/main/docs/rfcs/016-connection-routing.md#connection-url
+                            for cmdopt in value.split(',') {
+                                let nameval: Vec<&str> = cmdopt.split(':').collect();
                                 if nameval.len() == 2 {
                                     params.insert(nameval[0].to_string(), nameval[1].to_string());
+                                } else {
+                                    //todo: inform user / throw error message if options format is wrong.
```
Comment (hlinnaka) [inline libs/utils/src/pq_proto.rs:280]: yeah, should throw an error..

File: proxy/src/auth.rs [hunk 2]
```diff
@@ -6,7 +6,7 @@ use crate::config::{AuthBackendType, ProxyConfig};
 use crate::error::UserFacingError;
 use crate::stream::PqStream;
 use crate::{auth_backend, compute, waiters};
-use console::ConsoleAuthError::SniMissing;
+use console::ConsoleAuthError::SniMissingAndProjectNameMissing;
```
Comment (hlinnaka) [inline proxy/src/auth.rs:9]: I'd suggest just "ProjectNameMissing". We might have other ways of figuring out the project name in the future, and we don't want this to become "SniAndProjectNameAndBlahBlahMissing". At the end of the day, the error is that we couldn't figure out the project name, not from the SNI or the new option.

**Review by hlinnaka — COMMENTED** [review]
Replace 'cluster' with 'project' everywhere.

Is it still possible to provide Postgres options with the 'options' parameter? Normally, you can set server-side settings for the connection like this:
```
psql "postgres://localhost/postgres?options=--geqo%3Doff
```
Is that still possible? If you need to specify Neon settings, like the project, *and* Postgres server settings like `geqo=off`, what does it look like?

File: proxy/src/auth.rs (hunk 2 above)
Comment (funbringer) [reply proxy/src/auth.rs:9]: This is already known as [`ProjectNameError`](https://github.com/neondatabase/neon/blob/36281e3b475ac46570dd4f89a61fc525ff3f0a1c/proxy/src/auth/credentials.rs#L39). The patch should be rebased.

File: proxy/src/auth/credentials.rs [hunk 3]
```diff
@@ -37,10 +41,10 @@ impl ClientCredentials {
 
 #[derive(Debug, Error)]
 pub enum ProjectNameError {
-    #[error("SNI is missing, please upgrade the postgres client library")]
+    #[error("SNI info is missing. EITHER please upgrade the postgres client library OR pass the project name as a parameter: '..&options=project:<project name>..'.")]
```
Comment (funbringer) [inline proxy/src/auth/credentials.rs:44]: I'm not sure we should capitalize those words, but we have neither a formal procedure for figuring out texts like this nor a dedicated tech writer anyway, so I'm not insisting.

File: proxy/src/auth/credentials.rs [hunk 4]
```diff
@@ -68,11 +79,17 @@ impl TryFrom<HashMap<String, String>> for ClientCredentials {
 
         let user = get_param("user")?;
         let dbname = get_param("database")?;
+        let project_name = get_param("project");
+        let project_name = match project_name {
+            Ok(project_name) => Some(project_name),
+            Err(_) => None,
+        };
```
Comment (funbringer) [inline proxy/src/auth/credentials.rs:86]: See https://doc.rust-lang.org/std/result/enum.Result.html#method.ok

File: proxy/src/auth/credentials.rs [hunk 5]
```diff
@@ -26,6 +26,10 @@ pub struct ClientCredentials {
     // New console API requires SNI info to determine the cluster name.
     // Other Auth backends don't need it.
     pub sni_data: Option<String>,
+
+    // cluster_option is passed as argument from options from url.
+    // To be used to determine cluster name in case sni_data is missing.
+    pub project_name: Option<String>,
```
Comment (funbringer) [inline proxy/src/auth/credentials.rs:32]: Probably specifying both should be an error. What if they are incoherent?

File: proxy/src/auth/credentials.rs (hunk 3 above)
Comment (KlimentSerafimov) [reply proxy/src/auth/credentials.rs:44]: I capitalized them as 'command' words. I don't have a strong preference either.

File: proxy/src/auth/credentials.rs (hunk 4 above)
Comment (KlimentSerafimov) [reply proxy/src/auth/credentials.rs:86]: thanks, just learning this.

File: proxy/src/auth/credentials.rs (hunk 5 above)
Comment (KlimentSerafimov) [reply proxy/src/auth/credentials.rs:32]: Will add that check. Thanks!

File: proxy/src/auth/credentials.rs [hunk 6]
```diff
@@ -37,22 +42,44 @@ impl ClientCredentials {
...
 pub enum ProjectNameError {
-    #[error("SNI is missing, please upgrade the postgres client library")]
+    #[error("SNI info is missing. EITHER please upgrade the postgres client library OR pass the project name as a parameter: '...&options=project%3D<project-name>...'.")]
     Missing,
 
-    #[error("SNI is malformed")]
+    #[error("SNI is malformed.")]
     Bad,
+
+    #[error("Inconsistent project name inferred from SNI info and project option.")]
```
Comment (kelvich) [inline proxy/src/auth/credentials.rs:51]: I would also put exact two names that were given to highlight what was wrong

**Review by kelvich — APPROVED** [review]
Looks good, thanks. I also played locally with this branch a bit. Seems to be working as expected

**Pull request merged**